spec/main.spec.ts

```diff
@@ -106,11 +115,129 @@ describe('main', () => {
       expect(context.params).to.deep.equal(params);
       expect(context.resource.name).to.equal('ref/a/nested/b');
     });
+
+    it('should throw when snapshot path includes unspecified wildcards', () => {
+      const fft = new FirebaseFunctionsTest();
+      fft.init();
+
+      const snap = makeDataSnapshot(
+          'hello world',
+          '/ref/{wildcard}/nested/{anotherWildcard}',
+      );
+      const params = {
+          wildcard: 'at',
+          anotherWildcard: 'bat',
+      };
+      const wrapped = wrap(constructCF('google.firebase.database.ref.write'));
+      expect(() => wrapped(snap, { params })).to.throw();
+    });
+
+    it('should set auto-fill params based on DataSnapshot path', () => {
+      const fft = new FirebaseFunctionsTest();
+      fft.init();
+
+      const snap = makeDataSnapshot(
+        'hello world',
+        '/ref/cat/nested/that',
+      );
+
+      const params = {};
+      const wrapped = wrap(constructCF('google.firebase.database.ref.write'));
+      const result = wrapped(snap, { params });
+
+      expect(result.data._path).to.equal('ref/cat/nested/that');
+      expect(result.data.key).to.equal('that');
+      expect(result.context.params.wildcard).to.equal('cat');
+      expect(result.context.resource.name).to.equal('ref/cat/nested/that');
+    });
+
+    it('should error when DataSnapshot wildcard path does not match resource path', () => {
```

Comment:
This test checks the snapshot's provided path against the route in the function definition, I *think* that's still relevant because a dev could still path a snapshot which has a completely wrong path for a trigger, right?

Reply:
Ah, I see now, it is a poorly written test, but the intent is still useful, I think. I've changed it to..

```js
    it('should error when DataSnapshot path does not match trigger template', () => {
      const fft = new FirebaseFunctionsTest();
      fft.init();

      const snap = makeDataSnapshot(
          'hello world',
          '/different/at/nested/bat',
      );
      const wrapped = wrap(constructCF('google.firebase.database.ref.write'));
      expect(() => wrapped(snap, { params })).to.throw();
    });
  });
```

Does that seem more right?